synapse/lib/lmdbslab.py

```diff
@@ -166,6 +166,229 @@ def inc(self, name, valu=1):
+        self.abrvdb = slab.initdb(f'{name}:abrv2key')
+
+        self.prefix = prefix
+        if prefix is not None:
+            self.preflen = len(prefix)
+
+        self.offs = 0
+
+        item = self.slab.last(db=self.abrvdb)
+        if item is not None:
+            self.offs = s_common.int64un(item[0]) + 1
+
+    def getSubKeyVal(self, prefix):
+
+        if self.prefix is not None:
+            prefix = self.prefix + prefix
+
+        return SafeKeyVal(self.slab, self.name, prefix=prefix)
+
+    def get(self, name, defv=None):
+        '''
+        Get the value for a key.
+
+        Note:
+            This may only be used for keys < 503 characters in length. Use aget()
+            for keys above this limit or of unknown length.
+        '''
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        lkey = name.encode('utf8')
+
+        if len(lkey) > 502:
+            raise s_exc.BadArg('SafeKeyVal.get() may only be used with keys < 503 characters in length.')
+
+        if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+            return defv
+        return s_msgpack.un(byts)
+
+    async def aget(self, name, defv=None):
+        '''
+        Get the value for a key.
+
+        Note:
+            This may be used for keys of any length. The sync get() method may be
+            used for keys which have a known length < 503 characters.
+        '''
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        lkey = name.encode('utf8')
+
+        if len(lkey) <= 502:
+            if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+                return defv
+            return s_msgpack.un(byts)
+
+        for (abrv, realkey) in self.slab.scanByPref(lkey[:502], db=self.abrvdb):
+            if lkey == realkey:
+                if (byts := self.slab.get(lkey, db=self.valudb)) is None:
+                    return defv
+                return s_msgpack.un(byts)
+
+            await asyncio.sleep(0)
+
+        return defv
+
+    def set(self, name, valu):
+
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        byts = s_msgpack.en(valu)
+        name = name.encode('utf8')
+
+        if len(name) <= 502:
+            self.slab.put(name, byts, db=self.valudb)
+            return
+
+        abrv = s_common.int64en(self.offs)
+
+        self.offs += 1
+
+        lkey = name[:502] + abrv
+        self.slab.put(lkey, name, db=self.abrvdb)
+        self.slab.put(lkey, byts, db=self.valudb)
+
+        return valu
+
+    async def pop(self, name, defv=None):
+
+        if self.prefix is not None:
+            name = self.prefix + name
+
+        lkey = name.encode('utf8')
+
+        if len(lkey) <= 502:
+            if (byts := self.slab.pop(lkey, db=self.valudb)) is not None:
+                return s_msgpack.un(byts)
+            return defv
+
+        for (abrv, realkey) in self.slab.scanByPref(lkey[:502], db=self.abrvdb):
```

Comment:
Same Q here :) If so, we could probably make these sync assuming it's extremely unlikely to happen ( or we could limit keys sizes? )

Reply:
Yeah I considered making sync and async versions for a few more methods here because in some cases you have known length keys and don't have to worry about  > 502 byte keys. I was trying to avoid limiting key sizes because the length of the prefix used will eat into the allowed key length making the maximum length variable which seemed a bit clunky.